src/profile-logic/cpu.js

```diff
@@ -102,93 +101,50 @@ export function computeMaxCPUDeltaPerMs(profile: Profile): number {
-    const newSamples = {
-      ...samples,
-      threadCPUDelta,
-    };
-
-    const newThread = {
-      ...thread,
-      samples: newSamples,
-    };
-
-    return newThread;
+    return undefined;
   }
 
-  const newThreadCPUDelta: Array<number | null> = new Array(samples.length);
-  const cpuDeltaTimeUnitMultiplier = getCpuDeltaTimeUnitMultiplier(
-    sampleUnits.threadCPUDelta
-  );
-
-  for (let i = 0; i < samples.length; i++) {
-    // Ideally there shouldn't be any null values but that can happen if the
-    // back-end fails to get the CPU usage numbers from the operation system.
-    // In that case, try to find the closest number and use it to mitigate the
-    // weird graph renderings.
-    const threadCPUDeltaValue = findClosestNonNullValueToIdx(threadCPUDelta, i);
-
-    const threadCPUDeltaUnit = sampleUnits.threadCPUDelta;
-    switch (threadCPUDeltaUnit) {
-      // Check if the threadCPUDelta is more than the interval time and limit
-      // that number to the interval if it's bigger than that. This is mostly
-      // either a bug on the back-end or a bug on the operation system level.
-      // This happens mostly with µs values which is coming from macOS. We can
-      // remove that processing once we are sure that these numbers are reliable
-      // and this issue doesn't occur.
-      case 'µs':
-      case 'ns': {
-        const intervalInThreadCPUDeltaUnit =
-          i === 0
-            ? profileInterval * cpuDeltaTimeUnitMultiplier
-            : (samples.time[i] - samples.time[i - 1]) *
-              cpuDeltaTimeUnitMultiplier;
-        if (threadCPUDeltaValue > intervalInThreadCPUDeltaUnit) {
-          newThreadCPUDelta[i] = intervalInThreadCPUDeltaUnit;
-        } else {
-          newThreadCPUDelta[i] = threadCPUDeltaValue;
-        }
-        break;
-      }
-      case 'variable CPU cycles':
-        newThreadCPUDelta[i] = threadCPUDeltaValue;
-        break;
-      default:
-        throw assertExhaustiveCheck(
-          threadCPUDeltaUnit,
-          'Unhandled threadCPUDelta unit in the processing.'
-        );
+  const threadCPURatio: Float64Array = new Float64Array(threadCPUDelta.length);
+  threadCPURatio[0] = 0;
+
+  for (let i = 1; i < threadCPUDelta.length; i++) {
+    const referenceCpuDelta = maxThreadCPUDeltaPerMs * timeDeltas[i];
+    const cpuDelta = threadCPUDelta[i];
+    if (cpuDelta === null || referenceCpuDelta === 0) {
+      // Default to 100% CPU if the CPU delta isn't known or if no time has
+      // elapsed between samples.
+      // In profiles from Firefox, values at the beginning of threadCPUDelta can
+      // be null if the samples at the beginning were collected by the base
+      // profiler, which doesn't support collecting CPU delta information yet,
+      // see bug 1756519.
+      threadCPURatio[i] = 1;
+      continue;
     }
+
+    // Limit values to 1.0.
+    threadCPURatio[i] =
+      cpuDelta <= referenceCpuDelta ? cpuDelta / referenceCpuDelta : 1;
```

Comment:
Can it be larger than 1? Or is it mostly a defensive check?
Edit: Ah I see that in the tests we check if they are capped to 100%. That behavior was there already.

Reply:
Exactly - this matches the existing behavior for time-based CPU deltas, and the clamping does nothing in the "variable CPU cycles" case because we'll be dividing by the maximum value.